### PR TITLE
sql: Cluster-specific `Optimize` pipelines

### PR DESCRIPTION
**Pull request by aalexandrov**
Parser and planner changes required to support the 

```sql
CREATE CLUSTER ... FEATURES (...)
```

syntax proposed in #25010.

### Motivation

  * This PR adds a known-desirable feature.

Part of #25273.

### Tips for reviewer

The PR can be reviewed in three groups. See individual commit messages for details.

We first group defines a couple of macros (`optimizer_feature_flags` and `impl_simple_options`) that should minimize the amount of code changes needed as we add or remove optimizer feature flags. Unfortunately, this also required a change to the `waklabout` crate (cc @benesch I think that the extra complexity of teaching `walkabout` to simulate `impl_simple_options` is offset by the amount of time spent maintaining related parser structs in the future).

- expr: add `optimizer_feature_flags` macro
- sql-parser: add `impl_simple_options` macro
- walkabout: support `impl_simple_options!(...)` calls

The second group migrates the existing `EXPLAIN WITH(...)` options from a bespoke implementation to one that uses `impl_simple_options` and `generate_extracted_config` macro expansions. This is needed because it would allow us to standardize the steps required to add support for a new optimizer feature flag to both `EXPLAIN WITH(...)` and `CREATE CLUSTER ... FEATURES (...)`.

- sql-parser: use `generate_extracted_config!` for `ExplainConfig`
- explain: remove `TryFrom<BTreeSet<String>>` for `ExplainConfig`

The third group adds the SQL syntax extensions proposed in #25010 to the managed clusters. I'm not documenting them for the moment because they are not going to be publicly visible (see the `sql_bail!("FEATURES not supported for non-system users")` line).

- sql: add `CREATE CLUSTER ... FEATURES(...)` for managed clusters

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by aalexandrov** [comment]
@def- let me know if any of the nightly test failure look suspicious / new to this PR.

File: src/walkabout/src/parse.rs
Comment (def-) [inline src/walkabout/src/parse.rs:248]: Uncovered

File: src/walkabout/src/parse.rs
Comment (def-) [inline src/walkabout/src/parse.rs:130]: Uncovered

File: src/walkabout/src/parse.rs
Comment (def-) [inline src/walkabout/src/parse.rs:99]: Uncovered

File: src/repr/src/optimize.rs
Comment (def-) [inline src/repr/src/optimize.rs:45]: Uncovered

**Review by def- — COMMENTED** [review]
Nightly didn't show anything new: https://buildkite.com/materialize/nightlies/builds/6534
Coverage shows quite some unused code though, is that expected? https://buildkite.com/materialize/coverage/builds/359 (I marked them inline)

File: src/walkabout/src/parse.rs
Comment (aalexandrov) [reply src/walkabout/src/parse.rs:99]: See my [comment above](https://github.com/MaterializeInc/materialize/pull/25373#discussion_r1495419500).

File: src/walkabout/src/parse.rs
Comment (aalexandrov) [reply src/walkabout/src/parse.rs:248]: See my [comment here](https://github.com/MaterializeInc/materialize/pull/25373#discussion_r1495419500).

**Review by jkosh44 — APPROVED** [review]
Adapter changes LGTM

**Review by benesch — COMMENTED** [review]
Sorry for not turning the review of this around more quickly. @aalexandrov and I are discussing on Slack a way to move the macro's responsibilities to the `sql-parser` crate's build script. That will avoid needing to add a special case to the walkabout crate while retaining the benefits of automatically generating code.

Conversation here: https://materializeinc.slack.com/archives/C02FWJ94HME/p1708445928703659

**Comment by aalexandrov** [comment]
> @aalexandrov and I are discussing on Slack a way to move the macro's responsibilities to the `sql-parser` crate's build script. That will avoid needing to add a special case to the walkabout crate while retaining the benefits of automatically generating code.

I'm holding off with that until the SQL council decides whether it's OK to keep the `snake_case` format for feature flags and `EXPLAIN WITH(...)` options. If this is not the case, I will have to remove the `impl_simple_options` and move back to hand-rolled AST structs in the `sql-parser`, so the change proposed will no longer be needed.

**Comment by aalexandrov** [comment]
@benesch made the requested changes:

- Moved to fully spelled out individual tokens for the `EXPLAIN ... WITH (...)` and `CLUSTER ... FEATURES (...)` clauses. See the changes in the "sql-parser: add `impl_simple_options` macro (fixup)" commit.
- Moved the `mz-sql-parser`-specific extensions required for making `walkabout` happy to `sql-parser/build.rs`. See the updated contents of the "walkabout: support `impl_simple_options!(...)` calls" commit.

**Review by benesch — APPROVED** [review]
Thanks for iterating on this. I'm glad you were able to rearrange this so that the specialized logic can all live in the `sql-parser` crate.

I still think defining the `impl_simple_options!` macro as a Rust macro makes things more complicated than necessary, as the build script needs to parse the macro calls and conjure up the walkabout IR that corresponds to what the macro expands to.

It seems to me like it would be simpler to remove the macro and define the enums directly in the Rust source code. Then have the build script do what the macro is currently doing: load up the `*Options` enum variants (they'd be sitting in the walkabout IR) and spit out a Rust file with the various trait implementations for the enums that gets `include!`d into the crate just like the walkabout traits.

Maybe there's some reason that doesn't work out the way I hope, but it seems it would have the big benefit of there being only one place where the "macro" expansion happens, instead of the current situation, where the build script and the macro definition need to be kept perfectly in sync.

Anyway, not a blocking concern. Fine to leave as is and we can revisit in the future if it becomes an issue.

**Comment by aalexandrov** [comment]
The main issue at the moment is that 

> It seems to me like it would be simpler to remove the macro and define the enums directly in the Rust source code. Then have the build script do what the macro is currently doing: load up the `*Options` enum variants (they'd be sitting in the walkabout IR) and spit out a Rust file with the various trait implementations for the enums that gets `include!`d into the crate just like the walkabout traits.

I see... I did what you suggested. The code is probably a bit more, but probably also easier to understand and maintain.

**Comment by benesch** [comment]
> The code a bit more...

Oh, sorry, but you don't need to parse the module into walkabout IR twice! The IR already has exactly the information you need about the simple option enum variants. I pushed up a commit showing what I mean.

> Maybe I'll do one more pass to integrate `CodegenBuf` before merging this.

Since I was in the area I tackled this as well. It saved a handful of lines.

File: src/sql-parser/tests/testdata/explain
Comment (benesch) [inline src/sql-parser/tests/testdata/explain:36]: Something very weird happened to this file. Can we restore its original formatting?

File: test/sqllogictest/cluster_features.slt
Comment (benesch) [inline test/sqllogictest/cluster_features.slt:19]:
```suggestion
# Regular users cannot create clusters with FEATURES yet.
```

**Review by benesch — COMMENTED** [review]
Assuming you're on board with my latest two commits, LGTM, modulo the weird extra linebreaks in the testdata/explain file.

File: src/sql-parser/tests/testdata/explain
Comment (aalexandrov) [reply src/sql-parser/tests/testdata/explain]: My VSCode thinks that this is Markdown and does auto-format the file if I'm not careful.

**Comment by aalexandrov** [comment]
> > The code a bit more...
> 
> Oh, sorry, but you don't need to parse the module into walkabout IR twice! The IR already has exactly the information you need about the simple option enum variants. I pushed up a commit showing what I mean.
> > Maybe I'll do one more pass to integrate `CodegenBuf` before merging this.
> 
> Since I was in the area I tackled this as well. It saved a handful of lines.

Sadly I lost your suggestions because I did just that and then force-pushed my changes. I hope the final result is not too different from what you proposed. If this is the case feel free to send me the diff and I'll make the changes as part of the next PR.

**Comment by benesch** [comment]
> Sadly, I lost your suggestions because I did independently what you suggested and distributed the changes from my "fixup" commits before force-pushing to this branch (sadly GitHub doesn't seem to persist commits after a force-push). I hope that my final result is not too different from what you proposed (it amounts to ~100 LOC in `build.rs` which seems quite good). If I have missed something feel free to send me the diff and I'll make the changes as part of the next PR.

Looks quite similar. Here's the diff between our two approaches:

```diff
diff --git a/src/sql-parser/build.rs b/src/sql-parser/build.rs
index eec0a1bac0..f52805d834 100644
--- a/src/sql-parser/build.rs
+++ b/src/sql-parser/build.rs
@@ -13,15 +13,18 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::BTreeMap;
 use std::path::PathBuf;
 use std::{env, fs};
 
 use anyhow::{Context, Result};
+use mz_walkabout::ir::Item;
 
 const AST_DEFS_MOD: &str = "src/ast/defs.rs";
 
 fn main() -> Result<()> {
     let out_dir = PathBuf::from(env::var_os("OUT_DIR").context("Cannot read OUT_DIR env var")?);
+
     let ir = mz_walkabout::load(AST_DEFS_MOD)?;
 
     // Generate AST visitors.
@@ -33,13 +36,28 @@ fn main() -> Result<()> {
         fs::write(out_dir.join("visit.rs"), visit)?;
         fs::write(out_dir.join("visit_mut.rs"), visit_mut)?;
     }
+
     // Generate derived items for AST types modelling simple options.
     {
-        // Generate `Parser` methods.
-        let parse = simple_options::gen_parse(&ir);
+        // TODO: identify enums using an attribute
+        const ENUMS: &[&str] = &["ExplainPlanOptionName", "ClusterFeatureName"];
+
+        let mut enums = BTreeMap::new();
+        for name in ENUMS {
+            match ir.items.get(*name) {
+                None => panic!("enum {name} not found in IR"),
+                Some(Item::Struct(_) | Item::Abstract) => {
+                    panic!("type {name} is not an enum");
+                }
+                Some(Item::Enum(enum_)) => {
+                    enums.insert(*name, enum_);
+                }
+            }
+        }
+
+        let parse = simple_options::gen_parse(&enums);
+        let display = simple_options::gen_display(&enums);
         fs::write(out_dir.join("parse.simple_options.rs"), parse)?;
-        // Generate `AstDisplay` implementations.
-        let display = simple_options::gen_display(&ir);
         fs::write(out_dir.join("display.simple_options.rs"), display)?;
     }
 
@@ -47,68 +65,61 @@ fn main() -> Result<()> {
 }
 
 mod simple_options {
+    use std::collections::BTreeMap;
+
     use mz_ore::codegen::CodegenBuf;
-    use mz_walkabout::ir;
+    use mz_walkabout::ir::Enum;
 
-    // TODO: we might want to identify these enums using an attribute.
-    const SIMPLE_OPTION_ENUMS: [&str; 2] = ["ExplainPlanOptionName", "ClusterFeatureName"];
-
-    pub(super) fn gen_display(ir: &ir::Ir) -> String {
+    pub fn gen_display(enums: &BTreeMap<&str, &Enum>) -> String {
         let mut buf = CodegenBuf::new();
-        for enum_name in SIMPLE_OPTION_ENUMS {
-            if let Some(ir::Item::Enum(enum_item)) = ir.items.get(enum_name) {
-                write_ast_display(enum_name, enum_item, &mut buf)
-            }
+        for (name, enum_) in enums {
+            write_ast_display(name, enum_, &mut buf)
         }
         buf.into_string()
     }
 
-    fn write_ast_display(ident: &str, enum_item: &ir::Enum, buf: &mut CodegenBuf) {
-        let typ = ident.to_string();
-        let fn_fmt = "fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>)";
-        buf.write_block(format!("impl AstDisplay for {typ}"), |buf| {
-            buf.write_block(fn_fmt, |buf| {
+    fn write_ast_display(name: &str, enum_: &Enum, buf: &mut CodegenBuf) {
+        buf.write_block(format!("impl AstDisplay for {name}"), |buf| {
+            let fn_def = "fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>)";
+            buf.write_block(fn_def, |buf| {
                 buf.write_block("match self", |buf| {
-                    for v in enum_item.variants.iter().map(|v| &v.name) {
-                        let ts = separate_tokens(v, ' ').to_uppercase();
-                        buf.writeln(format!(r#"Self::{v} => f.write_str("{ts}"),"#));
+                    for v in enum_.variants.iter().map(|v| &v.name) {
+                        let ts = separate_tokens(&v, ' ').to_uppercase();
+                        buf.writeln(format!("Self::{v} => f.write_str(\"{ts}\"),"));
                     }
-                });
-            });
+                })
+            })
         });
-        buf.end_line();
     }
 
-    pub(super) fn gen_parse(ir: &ir::Ir) -> String {
+    pub fn gen_parse(enums: &BTreeMap<&str, &Enum>) -> String {
         let mut buf = CodegenBuf::new();
         buf.write_block("impl<'a> Parser<'a>", |buf| {
-            for enum_name in SIMPLE_OPTION_ENUMS {
-                if let Some(ir::Item::Enum(enum_item)) = ir.items.get(enum_name) {
-                    write_fn_parse(enum_name, enum_item, buf)
-                }
+            for (name, enum_) in enums {
+                write_fn_parse(name, enum_, buf)
             }
         });
         buf.into_string()
     }
 
-    fn write_fn_parse(ident: &str, enum_item: &ir::Enum, buf: &mut CodegenBuf) {
-        let typ = ident.to_string();
-        let msg = separate_tokens(&typ, ' ').to_lowercase();
-        let fn_name = format!("parse_{}", separate_tokens(&typ, '_').to_lowercase());
-        let fn_type = format!("Result<{typ}, ParserError>");
-        buf.write_block(format!("fn {fn_name}(&mut self) -> {fn_type}"), |buf| {
-            for v in enum_item.variants.iter().map(|v| &v.name) {
-                let kws = separate_tokens(v, ',').to_uppercase();
-                buf.write_block(format!("if self.parse_keywords(&[{kws}])"), |buf| {
-                    buf.writeln(format!("return Ok({typ}::{v})"));
-                });
+    fn write_fn_parse(name: &str, enum_: &Enum, buf: &mut CodegenBuf) {
+        let fn_ = separate_tokens(&name, '_').to_lowercase();
+        let msg = separate_tokens(&name, ' ').to_lowercase();
+
+        let fn_def = format!("fn parse_{fn_}(&mut self) -> Result<{name}, ParserError>");
+        buf.write_block(fn_def, |buf| {
+            for v in enum_.variants.iter().map(|v| &v.name) {
+                let keywords = separate_tokens(&v, ',').to_uppercase();
+                buf.write_block(format!("if self.parse_keywords(&[{keywords}])"), |buf| {
+                    buf.writeln(format!("return Ok({name}::{v})"));
+                })
             }
+
             buf.write_block("", |buf| {
-                buf.writeln(format!(r#"let msg = "a valid {msg}".to_string();"#));
-                buf.writeln(format!(r#"Err(self.error(self.peek_pos(), msg))"#));
-            });
+                buf.writeln(format!("let msg = \"a valid {msg}\".to_string();"));
+                buf.writeln(format!("Err(self.error(self.peek_pos(), msg))"));
+            })
         });
-        buf.end_line();
     }
 
     fn separate_tokens(name: &str, sep: char) -> String {
diff --git a/src/walkabout/src/ir.rs b/src/walkabout/src/ir.rs
index 701f65162a..914bd0b907 100644
--- a/src/walkabout/src/ir.rs
+++ b/src/walkabout/src/ir.rs
@@ -145,7 +145,7 @@ pub enum Type {
 /// This is a very, very lightweight semantic analysis phase for Rust code. Our
 /// main goal is to determine the type of each field of a struct or enum
 /// variant, so we know how to visit it. See [`Type`] for details.
-pub fn analyze(syn_items: &[syn::DeriveInput]) -> Result<Ir> {
+pub(crate) fn analyze(syn_items: &[syn::DeriveInput]) -> Result<Ir> {
     let mut items = BTreeMap::new();
     for syn_item in syn_items {
         let name = syn_item.ident.to_string();
diff --git a/src/walkabout/src/lib.rs b/src/walkabout/src/lib.rs
index c5c01d7781..003cf7c117 100644
--- a/src/walkabout/src/lib.rs
+++ b/src/walkabout/src/lib.rs
@@ -42,8 +42,6 @@ mod parse;
 pub mod ir;
 
 pub use gen::{gen_fold, gen_visit, gen_visit_mut};
-pub use ir::analyze;
-pub use parse::parse_mod;
 
 /// Loads type definitions from the specified module.
 ///
```

The differences to build.rs are immaterial. I would like to pull the changes to `walkabout` in, as they preserve the public API that is present on main. Can you apply those changes? I don't want to overwrite your changes by accident.

**Comment by aalexandrov** [comment]
The `mz_walkabout` changes and any changes to any `Cargo.toml` files leftovers from a prior version. I force-pushed a version that removes both.

**Review by benesch — APPROVED** [review]
Looks fantastic!

**Pull request merged**
